### ISSUE TITLE
[stage1] Haxelib replacement

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe-evolution#30]
# Foreword

Issues in this repositories tend to get bogged down in long-winded discussions without much of a result. We would like to try a slightly different approach which is outlined here: https://github.com/HaxeFoundation/haxe-evolution/wiki/Haxe-Projects

We are in Stage 1 here, which means that comments should be short and to the point as mentioned in the document. If you want to discuss the procedure, please open a separate issue.

Now on to the actual project.

<hr>

# Haxelib replacement

> Project lead: Simn
> Due date for stage 1: 2017-10-14 (2 weeks)

We are looking to replace haxelib with a better package manager solution.

Please share your ideas and concerns in a concise manner. Please also +1/-1 individual comments that you agree/disagree with.

**Comment by Simn** [comment]
The Haxe compiler should not have to call a separate tool to obtain the information required for compilation.

**Comment by kevinresol** [comment]
> The Haxe compiler should not have to call a separate tool to obtain the information required for compilation.

That means the compiler should be able to transform `-lib` tags into non`-lib` tags on its own, with the resolving logic being simple yet generic.

**Comment by elsassph** [comment]
Criteria:
- semver dependency management, and solution to handle conflicts as we can have only one version of a lib,
- global cache and local libs similarly to npm/yarn,
- support for git / file dependencies.

**Comment by damoebius** [comment]
NPM isn't enough ?

**Comment by ibilon** [comment]
The user should be able to use an alternative package manager (without making change or replacing the haxe exe).

**Comment by ousado** [comment]
The compiler should expose an interface for frontends like CLIs, IDEs, vscode-style language servers, and so on.

**Comment by ousado** [comment]
Whatever ends up as standard mechanism to invoke the compiler for one-shot compilations must be very fast, there should be no noticeable difference to invoking the compiler directly (in other words start-up times like that of nodejs or even the JVM are out of the question)

**Comment by ousado** [comment]
The standard mechanism to invoke the compiler will be a first point of contact for newcomers to the language. As such it should provide an outstanding user experience and documentation can't be an afterthought.

**Comment by ibilon** [comment]
We shouldn't change the current workflow and so not introduce any steps between changing a hxml and running it/doing a haxe command by hand.

**Comment by ibilon** [comment]
A migration plan is required:

* what happens to the already installed libraries? do they need to be reinstalled?
* do we keep the haxelib online repository as is? if not what about usernames?

**Comment by ousado** [comment]
Beyond building projects the standard mechanism to invoke the compiler should provide a standardized way to run other common tasks like tests, benchmarks, updating dependencies, etc. (much in the spirit of cargo).

**Comment by ibilon** [comment]
haxelib.json should be extended, proposed new fields (non exhaustive):
* repository field (to go with the current website field)
* listing the metas/defines of the library
* haxe version compatibility

**Comment by kevinresol** [comment]
> That means the compiler should be able to transform -lib tags into non-lib tags on its own, with the resolving logic being simple yet generic.


Because `-lib` is the only thing that cannot be "consumed directly" by the compiler.

**Comment by MangelMaxime** [comment]
- Support lock files so we have reproductive build system

**Comment by ibilon** [comment]
The following fields should be removed from haxelib.json (or its replacements) and dealt on the library repository website:
* description
* website
* contributors
* tags

since it doesn't make sense to update those by making new releases.

**Comment by ousado** [comment]
Interesting, what are the arguments against `tool test`, `tool bench`, `tool update` .. ?

**Comment by elsassph** [comment]
I know @damoebius' suggestion to use npm has been firmly downvoted but it isn't that silly - alternatively it would be interesting to have a npm-compatible haxelib API so we could be really interesting to have.

**Comment by EricBishton** [comment]
Configuration must be thoroughly documented and transparent.  On-disk storage such as hidden .dev, and .current files should be discouraged and instead replaced by a documented configuration file.

The final goal is to make it so that alternative library managers can use/manipulate the same library database without having to know the details of the previous manager.

**Comment by Simn** [comment]
> Interesting, what are the arguments against tool test, tool bench, tool update .. ?

Out of scope (except updates, but that's obvious). This is about a package manager, not a complete build tool, which would be one level higher.

**Comment by ousado** [comment]
> The final goal is to make it so that alternative library managers can use/manipulate the same library database without having to know the details of the previous manager.

That's too restrictive. While some package managers might agree on one database layout, others might not. A package manager using e.g. Nix as a backend will require a different layout than one based on a git repo layout, both of which would be reasonable choices to make. 

If what you mean is a common layout for a staging area, as well as for ex- and importing a package database to allow migration between package mangers, OTOH, that'd be a good idea.

**Comment by ousado** [comment]
> > Interesting, what are the arguments against tool test, tool bench, tool update .. ?

> Out of scope (except updates, but that's obvious). This is about a package manager, not a complete build tool, which would be one level higher.

I was under the impression we're talking about all the (future) components required for the workflows that haxelib is currently involved with, not just its role as a package manager.

If this is about package management only, I think we should look at the more principled approaches  - so basically read [A modular package manager architecture](https://dl.acm.org/citation.cfm?id=2401012), conclude that we're probably not going to come up with something better, and hence  "do whatever opam does".

**Comment by EricBishton** [comment]
> > The final goal is to make it so that alternative library managers can use/manipulate the same library database without having to know the details of the previous manager.

> That's too restrictive. While some package managers might agree on one database layout, others might not. A package manager using e.g. Nix as a backend will require a different layout than one based on a git repo layout, both of which would be reasonable choices to make.

It is and it isn't.  While I respect the decision of writers of the extended or alternative package managers to create the solutions they think most appropriate, I believe that the canonical package manager (the one delivered as part of the Haxe toolset) must set an open standard that other tools can inter-operate with.

**Comment by sangohan** [comment]
I think defining exactly what the role of the 'package manager' is (vs the build tool) must come first. 

Package manager is a tool that both downloads, and resolves package and its dependencies. It does not have to be part of a build tool. I feel that conflating both the build step, and the dependency resolution step into one tool will not work in the haxe ecosystem (since haxe is a source->source compiler, and each output target has its own canonical build tools). Therefore, to make a haxe package manager work for all targets, it must set about defining a format for defining dependencies of the current project being compiled (which is then in turn used by other projects). This format must include a specification for a way to retrieve its dependencies for each target. I think extending `haxelib.json` to include more data, including the retrieval mechanism for externally compiled libaries, is the way to go. 

`haxelib.json` :
``` 
{
  "dependencies": [
    {
        "haxe" :[
            { name: "pure-haxe-lib-1", version: "3.4.64"},
            { name: "my-custom-lib", location:"my-custom-lib/src"}
        ]
    },
    {
        "java" : [
            { name: "some-java-lib-1", location: "libs/java/some-java-lib-1.jar"},
            { name: "some-java-lib-2", location: "libs/java/some-java-lib-2.jar"},
        ]
    },
    {
        "js" : [
            { name: "some-js-lib-1", location: "libs/node_modules/some-js-lib-1", hxml:"-D 'some_stuff'"},
            { name: "some-js-lib-2", location: "libs/node_modules/some-js-lib-2"},
        ]
    }
  ]
}
```
This dependency will be hand written by the author of the haxe library/project. It is then up to the author of the project to work out how to put the required external dependency files in to the locations described by this file. `haxelib.exe` will parse this dependency description, and in the case of pure haxe dependencies, resolve transitives, and output a `hxml` file required to compile the current project.

I am deliberately not integrating any existing external dependency tool knowledge into the haxe package management, because any target output language is subject to change, and haxe is not able to continuously keep up. The authors of libraries need to figure out the best way to integrate it into the target language's ecosystem.

**Comment by jcward** [comment]
> This format must include a specification for a way to retrieve its dependencies for **each target**.

Woah, woah, woah. Aren't we just talking about Haxe dependencies? Is there any merit in extending to target-specific (java/js/etc) external libs? Isn't that really the job of a build tool?

**Comment by jcward** [comment]
You know, the only thing really odd with haxelib today is that you specify versions globally for all projects (`haxelib set hxfoo 1.2.3`), instead of at the individual project level (e.g. in hxml, `-lib hxfoo 1.2.3`). Let's not forget that haxelib is:

- easy for users out of the box
- easy to contribute to
- has git support
- has a decent website
- user can build his own version-aware tooling around it (e.g. setup versions before each build.)

Consider this:

- The very idea that a project should be able to specify a versioned library sounds great, but it leads you down a complex road.
- If projects must specify dependencies, so must libs also.
- Then you need transitive dependency resolution.
- Then you need to deal with conflicts, perhaps providing options for overrides, version ranges, etc (e.g. the apt packaging system has keywords like depends, provides, recommends, suggests, replaces, breaks.)
- Assuming symver gives us some leeway in this (e.g. we can assume all `1.y.z` libraries are indeed compatible), you then assume third party authors know and respect symver. What about mistakes? You then need a blacklist.
- How do git or local versions inter-operate with symver?

All this to say: managing versioned dependencies is complicated. And people have opinions based on their history (I like how apt works. I like how npm works. I like ...) Again, because haxelib today outsources version management to the user, we will by definition arrive at a system that's more complicated than haxelib today. But will it be more usable? Or better? Does it really belong in the compiler? I'd say it's unclear.

TLDR: I suppose I'm just skeptical until we see an actual proposal.

**Comment by Simn** [comment]
@sangohan @jcward: I removed your comments because they are way too wordy/not actionable. I've pasted them here so they are not lost: https://gist.github.com/Simn/2b13809d830cfd16dd5ba2f9c97cac6b

**Comment by Matan** [comment]
- Maintain a local cache of downloaded libs.
- Symlink cached libs to projects to save space. (flag to control behaviour and/or force local install)
- Support "globally" installing libs.
- Support `licenses` sub method to output all dependency licenses much like PHP's `composer licenses`.

**Comment by kiroukou** [comment]
I recently discovered https://lib.haxe.org/p/hmm  library which does a pretty nice job on that part with current limitations.

For me, the current haxelib version isn't really far from what I need, but here is my wishlist : 

- Having a way to abstract the lib source  (git, svn, local, ...)
- Having 2 scopes of lib loading : global and local
- Recursive loading (a lib depending itself on an other, should be able to load it automatically).  That leads to a version conflict resolver !
- Having/Keeping a comprehensive synthax

**Comment by jasononeil** [comment]
It should be possible to open a directory that a project is in, and when I type `haxe build.hxml` it automatically uses the configured version of Haxe for the project, and the right set of haxelibs/dependencies.

(This one might be controversial but I thought I'd throw it in!)

**Comment by jasononeil** [comment]
If I install a new dependency, upgrade a dependency, or change my dependencies in any way, this should show up in a way that is tracked in a file in my version control, and can be used to restore the same state on a different computer.

(Similar to a `yarn.lock` or `haxe_libraries/*.hxml` in haxeshim/lix)

**Comment by clemos** [comment]
I think we should deprecate `haxelib run` to avoid depending on neko.

**Comment by Simn** [comment]
> I think we should deprecate haxelib run to avoid depending on neko.

Don't have to deprecate it, but it shouldn't require a run.n. It could, by convention, make haxe interpret a Run.hx script instead.

**Comment by ibilon** [comment]
> (Similar to a yarn.lock or haxe_libraries/*.hxml in haxeshim/lix)

Can't put that in version control since it contains computer specific path into it.

As an alternative to a new file (like yarn.lock) maybe lock the versions directly in the hxml?

**Comment by jasononeil** [comment]
@ibilon here is how lix deals with it, no absolute paths:

```
# @install: lix --silent download "haxelib:hxnodejs#4.0.9" into hxnodejs/4.0.9/haxelib
-D hxnodejs=4.0.9
-cp ${HAXESHIM_LIBCACHE}/hxnodejs/4.0.9/haxelib/src
-D nodejs
--macro allowPackage('sys')
--macro _hxnodejs.VersionWarning.include()
```

Here is how yarn.lock deals with it (basically, it doesn't, it just uses relative paths):

```
angular-cookies@^1.6.4:
  version "1.6.4"
  resolved "https://registry.yarnpkg.com/angular-cookies/-/angular-cookies-1.6.4.tgz#c28f3f6aac7a9826c1e45f1d6807240036e5b26d"
```

**Comment by back2dos** [comment]
The package manager should use secure protocols (HTTPS vs. HTTP) for downloading packages (I found this to place limitations on the choice of a suitable runtime).

**Comment by markknol** [comment]
From Haxe code it should still be possible to know which libs are used. so doing `#if (libname > 4)`. It would also be useful to be able to know which classes come which library (in macros/ documentation xmls/dependency dumps etc)

**Comment by jgranick** [comment]
I have many feelings on haxelib, but quickly l want to share some "wish list" items:

 * Support for multiple remote repositories
 * Support for cascading local library paths
 * Support for wildcard versions (IE `-lib lime:3.1.*`)
 * Support for installing shortcut commands for tools

Multiple remote repositories would allow companies to generate their own internal versions and distribute more easily, or enable "beta" or "dev" channel repositories.

Maintaining class path directory is not reasonable for dealing with source files. Similarly, it would be nice to be able to support multiple library paths locally, like how class paths are resolved.

Support for wildcard versions would really, really help (especially if semver is respected). You can say you require library version `1.0.*`, allowing patch releases. This is a great middle-solution between "I need version 1.0.4 exactly" and "just give me any version at all"

As a cherry on top, it would be nice if libraries could install command aliases, such as `lime` for `haxelib run lime` or `as3hx` for `haxelib run as3hx`

**Comment by elsassph** [comment]
@clemos `haxelib run` isn't a problem per se, but it should be extended to other runtimes, and be configurable in `haxelib.json`.

**Comment by elsassph** [comment]
Caching libs is good, using symlinks is bad: someone could mistakenly modify a file and it would affect the cache / other projects.
Libs should be copied locally in the project (unless optional CI optimisation to use cache).

**Comment by clemos** [comment]
@Simn @elsassph I agree it's not so much about about deprecating `haxelib run` itself, but rather about deprecating `run.n` in favor of `Run.hx` ;)

**Comment by bablukid** [comment]
The package manager should allow custom fields in haxelib.json 
It would be very usefull for librairies which are tied to a specific framework or runtime.
i.e : For externs, we could provide a link to the original library and infos about how to integrate it in your haxe project
i.e : a web framework which loads plugins (haxelibs) that have various assets  ( like models, controllers, templates and translation files )

**Comment by andyli** [comment]
I would like to point out that it is important to consider how to handle "native code" (could be compiled ndll, C/C++/JS/... compiled libraries or source files).

One reason is that it is wasteful to download binaries of all architectures. e.g. Lime's `ndll` folder is 108M, but if you're only using Windows, you actually only need the 5.1M `ndll/Windows` folder.

We should also have some guidelines regarding to how to store the info of those "native" things, e.g. where is the original source, what param is used when building (use of -PIC/-O3/debug info), any additional dependencies that is not bundled etc.

**Comment by kevinresol** [comment]
Build tools (hxcpp,hxjava,hxcs) should not rely on haxelib too. 
They can be invoked by Context.onAfterGenerate or haxe --run

**Comment by Simn** [comment]
Stage 1 has concluded, I'll prepare the discussion part for Monday.
